spec/thread_safety_spec.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+RSpec.describe "thread safety" do # rubocop:disable RSpec/DescribeClass
+  context "when two threads accessing unmemoized zero-args method" do
+    # Tip of the hat to @jeremyevans,
+    # see https://github.com/panorama-ed/memo_wise/pull/224
+    let(:thread_return_values) do
```

Comment:
I think this whole block could be simplified to

```ruby
let(:thread_return_values) do
  instance = class_with_memo.new

  Array.new(2) do
    Thread.new { instance.return_first_thread_id }
  end.map(&:value)
end
```

Rationale:
* Now that it's deterministic, we don't need the `1000.times` block
* `Array#new` takes an optional block, so we don't need the `#map` there too
* We don't need the break condition anymore
* No need to define `threads` or `thread_return_values`

Reply:
@jemmaissroff while it is true that in the current code and test, we can indeed reliably reproduce the race on the first iteration, are you sure it's best to leave that assumption in place for an on-going regression test? In general, when leaving tests for previous race conditions in code, my norm has been to leave enough iterations up to 1 or 2 seconds of test time, so that if we every have a case that occurs probabilistically, which is the norm, we will eventually catch it in CI.

I'm open to thoughts either way! I just don't want my "success" here in triggering on the first iteration to have too much influence in how we think about designing race condition regression tests, if that makes sense.

Reply:
I don't think I have strong feelings here. I see the argument for letting this run for a while to catch nondeterministic failures, but I also think @jemmaissroff 's point about simplifying this if the failure _is_ deterministic makes sense. I'd be happy with either:

- simplifying the test as @jemmaissroff suggests
- keeping the code as-is and adding a comment explaining the rationale that @ms-ati outlined above

Reply:
Request for re-review @jemmaissroff -- I've extracted code and added comments to provide a rationale for time-limited probabilistic testing strategy for race conditions. What do you think?

Reply:
Since this race condition reproduction is deterministic, I personally was confused by the repetition here.

Although I can admire the elegance of `check_repeatedly`, it seems to me to be over-engineering in this case and solving for a problem that we don't currently have. If we at some future point either have a race condition which we need to repeatedly run to reproduce, or this race condition for some reason requires repetition, this method looks perfect.

In the absence of this however, I think the repetition here overcomplicates and makes it harder to reason about a test which only requires one iteration to accomplish its goal.

Reply:
@jemmaissroff Ok, it seems like you would advocate for not having the practice for time-limited probabilistic testing in place until we have discovered a race condition which manifests probabilitically?

I thought having this in place in this new file `thread_safety.rb` would lower the work required for us to add tests for other race conditions, as we work to land thread-safety and documentation of thread contention behavior across the project.

So is there a natural tension between these interests?

1. Avoid engineering test code for probabilistic race conditions until we have one at hand
2. Make it easier to detect and add tests for probabilistic race conditions

Reply:
I might be missing context here - do we have plans to add any other race condition tests? Or to do more work on creating more race condition tests?

Reply:
_To be clear: I am happy removing this machinery for now!_

I am hoping to reach alignment on whether there are distinct interests at play in this decision? Are we choosing to serve 1 over 2, or are saying that 2 is not a valid interest ("over-engineering" and "overcomplicates")?

Reply:
Yes, I am aligned with removing it.

I'm trying to understand whether 2 is a valid interest. My assumption earlier in this thread was that we don't have plans to add other race condition tests, and so under that assumption I suggested that it is overcomplicating our codebase. When you outlined the interests, it made me realize I had made this assumption, and so I'm looking for clarity on whether we do have plans to add more race condition tests or not. 

If we don't have plans to do this, I would say yes, accommodating a case that doesn't exist is not a valid interest.

If we do have plans to do this, I'm interested in hearing more about this before I feel equipped to judge which interest we should prioritize.

Reply:
Ah ok, thanks for clarifying @jemmaissroff. I had been under the assumption that this was the beginning of a series of work to eventually fully answer #184 -- and I had been imagining adding test coverage for potential races across all of the memoization code paths.

But I find it very persuasive that such tooling can wait at least one more PR, since this is the very first, none of those other plans might happen, and it isn't needed for just this PR.

I'll remove it for now and we can revisit if we do end up moving forward with thread-safety tests across all of the memoization paths.

Reply:
@jemmaissroff Removed as discussed, please re-review?

Reply:
Well 😆 the joke is on us @jemmaissroff @JacobEvelyn -- in fact, it appears that the issue is not actually deterministic, just appears often. So in fact, trying it once is not enough, as you can see from these CI failures: https://github.com/panorama-ed/memo_wise/runs/4231683800?check_suite_focus=true

Q: Shall I return the machinery for time-limited probabilistic tests then?